### PR TITLE
Make time axis

### PR DESCRIPTION
**Pull request by alexkjames**
Addresses #330

### CODE REVIEW

**Comment by alexkjames** [comment]
Still to do:

*Update gkernel to align behavior
*Update common time to align behavior

*Test methods and make sure defaults are all the same
*Test methods to make sure everything works as advertised

**Comment by CommonClimate** [comment]
> Test methods and make sure defaults are all the same
defaults should be the same for gkernel and bin. I was under the impression that, once interp() can properly handle NaNs, then we should allow users to specify finer time resolutions, in a way that would trip up the other two methods.  Am I mis-remembering?

**Comment by alexkjames** [comment]
@CommonClimate You're mostly correct. However, all of the methods will be allowed to specify finer time resolutions. The defaults for bin and gkernel will be to avoid this behavior, though the user will be free to do these things if they so wish. So 'make all defaults are the same' is a bit misleading on my part, I think that was the product of friday brain fog. I meant it should be that if you pass in the same arguments to each function, you'll receive the same time axis. The value axis will ideally be the main thing that varies between each method. If this is out of line with your expectations I'm happy to discuss further, I was under the impression that one of the main reasons we were overhauling these functions was to put them on similar semantic ground

**Comment by CommonClimate** [comment]
Yes, we are in agreement ; I was just thrown off by 'make all defaults are the same' .  The function will indeed be common to all regridding methods, and I was envisioning that it would be called within the API level functions `bin`, `gkernel`, `interp` with defaults set at the level of those functions. Once those behave as they should on realistic series, then I expect `common_time` will be greatly simplified. 

BTW, I am thinking that instead of `make_time_axis` we should actually call this utility function `make_even_time`, because it will be even, correct? If so, that is an important keyword to put in.   I suggest to set no defaults at all for this function at the utils level, so that we make it clear that it is the job of the API-level functions that call it to set appropriate defaults. However, if you see a reason to have universal defaults, let me know.

**Comment by alexkjames** [comment]
I see, currently most of the work done for each of these functions is done within the utilities, so just to be explicit we would be changing that to having the time axis step done at the API level right? Currently I have this being done within the utilities as well, though it should be straightforward to change.

**Comment by CommonClimate** [comment]
Yes, done within the utils-level function, but the setting of the step should be done at the API level, because it will vary between those regridding types (.e.g interp will use a different step style than bin). Do `start` and `stop` need to vary as well? Perhaps not, so I suggest having them set as defaults in the utils-level function, and only override them if need be.

**Comment by CommonClimate** [comment]
The key is to have all the common bits handled by the same piece of code, to minimize unwanted variation. But some variations are desirable, and should be made explicit!

**Comment by alexkjames** [comment]
Well put! One minor question that has risen with regards to how start,step and stop are interpreted is with regards to the difference between interp and bin. For interp these arguments are used to specify the time axis which will serve as the basis for interpolation. This usage is also possible for bin and gkernel, but requires "half bins" at the beginning and end of the time axis, as the bin edges must be set as the midpoints between times, but should not (in my opinion) extend beyond the start and stop times. For example, if start = 0, stop 3, and step (or bin_size) = 1, the time axis = [0,1,2,3], but the bins are [0,.5,1.5,2.5,3]. The alternative is to have the "time axis" in this case just define the bin edges, in which case, using the above example, bin_edges = [0,1,2,3] and the time_axis = [.5,1.5,2.5]. I think the latter makes more sense from a contextless perspective, but it does mean that the arguments are interpreted differently between the functions. Do you have any thoughts? @khider might also have an opinion. 

If my description just elicits more confusion, we can also save this discussion for the dev meeting tomorrow.

**Comment by CommonClimate** [comment]
Thanks for that. This is a clear indication that the very meaning of "time axis" varies from one application to the next, particularly this issue of edges vs midpoints. We do want to be very explicit about that, and maybe there should be two separate functions (one for bins, one for interpolation times)? Perhaps not, but worth thinking about. As long as the doc is clear, I think it's OK to have one function that does different things depending on context (edges vs midpoints), but you'll find out when implementing/testing whether that's the right approach!

File: pyleoclim/utils/tsutils.py [hunk 1]
```diff
@@ -136,6 +150,13 @@ def bin(x, y, bin_size=None, start=None, stop=None, evenly_spaced = True, statis
...
         The standard error on the mean in each bin
 
+    Notes
+    -----
+
+    `start`, `stop`, `bin_size`, and `step_style` are interpreted as defining the `bin_edges` for this function.
+    This differs from the `interp` interpretation, which uses these to define the time axis over which interpolation is applied.
+    For `bin`, the time axis will be specified as the midpoints between `bin_edges`, unless `time_axis` is explicitly passed.
+
     See also
```
Comment (CommonClimate) [inline pyleoclim/utils/tsutils.py:160]: THis is where you put the reference to scipy.stats.binned_statistic (with URL)

**Comment by CommonClimate** [comment]
NB3 looks like it works, so I'll merge.

**Pull request merged**